### PR TITLE
Adding pull inactivity timeout

### PR DESCRIPTION
**Pull request by petderek**
### Summary
* Carrying #1566 
* Changelog fix (since we dropped 1.20.4)
* longer timeout defaults on windows. 

### Implementation details

Windows containers tend to be larger, and so we see more failures due to extraction timeouts on windows instances. For linux, this doesn't need to be larger except in a smaller subset of scenarios.

This problem boils down to an issue with the Inactivity Timeout feature of go-dockerclient. For the image create case, docker doesn't separate the 'pull' from the 'extract' workflow, so the timeout applies to both. I've added a comment in the moby branch in #1606 

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [ ] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [ ] Integration tests on Linux (`make run-integ-tests`) pass
- [ ] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [ ] Functional tests on Linux (`make run-functional-tests`) pass
- [ ] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: yes

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -131,6 +131,7 @@ additional details on each available environment variable.
 | `ECS_ENABLE_CPU_UNBOUNDED_WINDOWS_WORKAROUND` | `true` | When `true`, ECS will allow CPU unbounded(CPU=`0`) tasks to run along with CPU bounded tasks in Windows. | Not applicable | `false` |
 | `ECS_TASK_METADATA_RPS_LIMIT` | `100,150` | Comma separated integer values for steady state and burst throttle limits for task metadata endpoint | `40,60` | `40,60` |
 | `ECS_SHARED_VOLUME_MATCH_FULL_CONFIG` | `true` | When `true`, ECS Agent will compare name, driver options, and labels to make sure volumes are identical. When `false`, Agent will short circuit shared volume comparison if the names match. This is the default Docker behavior. If a volume is shared across instances, this should be set to `false`. | `false` | `false`|
+| `ECS_DOCKER_PULL_INACTIVITY_TIMEOUT` | 1m | The time to wait after docker pulls complete waiting for extraction of a container. Useful for tuning large Windows containers. | 1m | 3m |
```
Comment (sharanyad) [inline README.md]: Wouldn't this be more descriptive if it was `ECS_IMAGE_PULL_INACTIVITY_TIMEOUT`?

File: agent/config/config.go [hunk 2]
```diff
@@ -77,6 +77,10 @@ const (
 	// a task's container. This is used to enforce sane values for the config.TaskCleanupWaitDuration field.
 	minimumTaskCleanupWaitDuration = 1 * time.Minute
 
+	// minimumDockerPullInactivityTimeout specifies the minimum duration to wait before cleaning up
```
Comment (sharanyad) [inline agent/config/config.go]: this comment needs update

File: agent/config/config_test.go [hunk 3]
```diff
@@ -269,6 +290,26 @@ func TestInvalidValueContainerStartTimeout(t *testing.T) {
...
+	defer setTestRegion()()
+	defer setTestEnv("ECS_DOCKER_PULL_INACTIVITY_TIMEOUT", "0s")()
+	ctrl := gomock.NewController(t)
+	mockEc2Metadata := mock_ec2.NewMockEC2MetadataClient(ctrl)
+	conf, err := NewConfig(mockEc2Metadata)
+	assert.NoError(t, err)
+	assert.Equal(t, conf.DockerPullInactivityTimeout, defaultDockerPullInactivityTimeout, "Wrong value for DockerPullInactivityTimeout")
+}
+
+func TestInvalidValueDockerPullInactivityTimeout(t *testing.T) {
```
Comment (sharanyad) [inline agent/config/config_test.go]: how is this different from the test `TestInvalidValueDockerInactivityTimeout`?

File: README.md (hunk 1 above)
Comment (petderek) [reply README.md]: Either  Docker / Image -- doesn't make a difference to me here. Its pulling an image, but the timeout is also specific to a docker operation (pull and unpack).

File: agent/config/config_test.go (hunk 3 above)
Comment (petderek) [reply agent/config/config_test.go]: It's a different input edge case test that follows the same setup / teardown and has the same result.  Its useful -- makes sure we don't accidentally handle the zero case incorrectly in the future.

File: README.md (hunk 1 above)
Comment (sharanyad) [reply README.md]: I would choose Image because the variable can still be used if we move away from Docker :)

File: agent/config/config_test.go [hunk 4]
```diff
@@ -281,7 +281,7 @@ func TestInvalidValueDockerPullInactivityTimeout(t *testing.T) {
 	defer setTestEnv("ECS_DOCKER_PULL_INACTIVITY_TIMEOUT", "-10s")()
 	conf, err := NewConfig(ec2.NewBlackholeEC2MetadataClient())
 	assert.NoError(t, err)
-	assert.Equal(t, conf.ImagePullInactivityTimeout, minimumImagePullInactivityTimeout, "Wrong value for ImagePullInactivityTimeout")
+	assert.Equal(t, conf.ImagePullInactivityTimeout, defaultImagePullInactivityTimeout, "Wrong value for ImagePullInactivityTimeout")
```
[the remark this answers is not in the document]
Comment (sharanyad) [reply agent/config/config_test.go]: Shouldn't both be `minimumImagePullInactivityTimeout `?

[the remark this answers is not in the document]
Comment (haikuoliu) [reply agent/config/config_test.go]: To make it clear, I copied two tests on commit `7c2d5e` below:

```
func TestInvalidValueDockerInactivityTimeout(t *testing.T) {
	defer setTestRegion()()
	defer setTestEnv("ECS_IMAGE_PULL_INACTIVITY_TIMEOUT", "-10s")()
	conf, err := NewConfig(ec2.NewBlackholeEC2MetadataClient())
	assert.NoError(t, err)
	assert.Equal(t, conf.ImagePullInactivityTimeout, minimumImagePullInactivityTimeout, "Wrong value for ImagePullInactivityTimeout")
}
```


```
func TestInvalidValueDockerPullInactivityTimeout(t *testing.T) {
	defer setTestRegion()()
	defer setTestEnv("ECS_DOCKER_PULL_INACTIVITY_TIMEOUT", "-10s")()
	conf, err := NewConfig(ec2.NewBlackholeEC2MetadataClient())
	assert.NoError(t, err)
	assert.Equal(t, conf.ImagePullInactivityTimeout, defaultImagePullInactivityTimeout, "Wrong value for ImagePullInactivityTimeout")
}
```

**Comment by wattdave** [comment]
LGTM; should I close mine?

File: agent/config/config_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (petderek) [reply agent/config/config_test.go]: oh interesting! I'll look at this.

**Comment by petderek** [comment]
@wattdave Sure! Thanks much for contributing. 😄
